core/consensus/common/src/import_queue.rs

```diff
@@ -97,227 +96,445 @@ pub trait ImportQueue<B: BlockT>: Send + Sync {
+				};
+				for msg in port.iter() {
+					// Working until all senders have been dropped...
+					match msg {
+						BlockImportWorkerMsg::ImportBlocks(origin, blocks) => {
+							worker.import_a_batch_of_blocks(origin, blocks)
+						}
+						_ => unreachable!("Import Worker does not receive the Imported message"),
+					}
+				}
+			})
+			.expect("ImportQueueWorker thread spawning failed");
+		sender
+	}
+
+	fn import_a_batch_of_blocks(&self, origin: BlockOrigin, blocks: Vec<IncomingBlock<B>>) {
+		let count = blocks.len();
+		let mut imported = 0;
+
+		let blocks_range = match (
+			blocks
+				.first()
+				.and_then(|b| b.header.as_ref().map(|h| h.number())),
+			blocks
+				.last()
+				.and_then(|b| b.header.as_ref().map(|h| h.number())),
+		) {
+			(Some(first), Some(last)) if first != last => format!(" ({}..{})", first, last),
+			(Some(first), Some(_)) => format!(" ({})", first),
+			_ => Default::default(),
+		};
+
+		trace!(target:"sync", "Starting import of {} blocks {}", count, blocks_range);
+
+		let mut results = vec![];
+
+		let mut has_error = false;
+
+		// Blocks in the response/drain should be in ascending order.
+		for block in blocks {
+			let import_result = if has_error {
+				Err(BlockImportError::Error)
+			} else {
+   				import_single_block(
+   					&*self.block_import,
+   					origin.clone(),
+   					block.clone(),
+   					self.verifier.clone(),
+   				)
+			};
+			let was_ok = import_result.is_ok();
+			results.push((import_result, block.hash));
+			if was_ok {
+				imported = imported + 1;
+			} else {
+				has_error = true;
+			}
+		}
+
+		let _ = self
+			.result_sender
+			.send(BlockImportWorkerMsg::Imported(results));
+
+		trace!(target: "sync", "Imported {} of {}", imported, count);
+	}
 }
 
 /// Hooks that the verification queue can use to influence the synchronization
 /// algorithm.
 pub trait Link<B: BlockT>: Send {
 	/// Block imported.
 	fn block_imported(&self, _hash: &B::Hash, _number: NumberFor<B>) { }
+	/// Justification import result.
+	fn justification_imported(&self, _who: Origin, _hash: &B::Hash, _number: NumberFor<B>, _success: bool) { }
 	/// Request a justification for the given block.
 	fn request_justification(&self, _hash: &B::Hash, _number: NumberFor<B>) { }
 	/// Maintain sync.
-	fn maintain_sync(&self) { }
+	fn maintain_sync(&self) {}
 	/// Disconnect from peer.
-	fn useless_peer(&self, _who: Origin, _reason: &str) { }
+	fn useless_peer(&self, _who: Origin, _reason: &str) {}
 	/// Disconnect from peer and restart sync.
-	fn note_useless_and_restart_sync(&self, _who: Origin, _reason: &str) { }
+	fn note_useless_and_restart_sync(&self, _who: Origin, _reason: &str) {}
 	/// Restart sync.
-	fn restart(&self) { }
+	fn restart(&self) {}
 }
 
 /// Block import successful result.
 #[derive(Debug, PartialEq)]
-pub enum BlockImportResult<H: ::std::fmt::Debug + PartialEq, N: ::std::fmt::Debug + PartialEq> {
+pub enum BlockImportResult<N: ::std::fmt::Debug + PartialEq> {
```

Comment:
I couldn't quite remember why I did this, but they're actually a reason for it :) 

In the previous code, the `import-thread` the "handling of results" and the "importing" where all done by the "import-thread". Now the worker does the importing, and the importer handles the result(and handles request such as "is importing"). 

Note that `BlockImportResult` is only used in the `Ok` case, in the `Err` case `BlockImportError` is used. The error doesn't contain the hash. This previously didn't matter be because all the logic happened in the same thread/function. For example in order to remove the hashes from the queue of block to be imported, they would be stored in a variable here

https://github.com/paritytech/substrate/blob/c188a3dc1505167748121a8f6cb9e970ff47df91/core/consensus/common/src/import_queue.rs#L275

and handled after the importing was done at

https://github.com/paritytech/substrate/blob/c188a3dc1505167748121a8f6cb9e970ff47df91/core/consensus/common/src/import_queue.rs#L288

So I actually removed the hash from `BlockImportResult`, but each "result" message that comes in from the worker now contains a tuple of (hash, result), which is necessary because we need the hash also when the result in an error...

Reply:
Yeah I know the same behavior is being kept, I just meant from a semantic point of view. Also no strong opinion on this, fine as is.